rust/src/dataset/optimize.rs

```diff
@@ -0,0 +1,779 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Table maintenance for optimizing table layout.
+//!
+//! As a table is updated, it's layout can become suboptimal. For example, if
+//! a series of small streaming appends are performed, eventually there will be
+//! a large number of small files. This imposes an overhead to track the large
+//! number of files and for very small files can make it harder to read data
+//! efficiently. In this case, files can be compacted into fewer larger files.
+//!
+//! To compact files in a table, use the [compact_files] method. This currently
+//! can compact in two cases:
+//!
+//! 1. If a fragment has fewer rows than the target number of rows per fragment.
+//! 2. If a fragment has a higher percentage of deleted rows than the provided
+//!    threshold.
+//!
+//! ```rust
+//! # use std::sync::Arc;
+//! # use tokio::runtime::Runtime;
+//! # use arrow_array::{RecordBatch, RecordBatchIterator, Int64Array};
+//! # use arrow_schema::{Schema, Field, DataType};
+//! use lance::{dataset::WriteParams, Dataset, dataset::optimize::compact_files};
+//!
+//! # let mut rt = Runtime::new().unwrap();
+//! # rt.block_on(async {
+//! #
+//! # let test_dir = tempfile::tempdir().unwrap();
+//! # let uri = test_dir.path().to_str().unwrap().to_string();
+//! let schema = Arc::new(Schema::new(vec![Field::new("test", DataType::Int64, false)]));
+//! let data = RecordBatch::try_new(
+//!     schema.clone(),
+//!     vec![Arc::new(Int64Array::from_iter_values(0..10_000))]
+//! ).unwrap();
+//! let reader = RecordBatchIterator::new(vec![Ok(data)], schema);
+//!
+//! // Write 100 small files
+//! let write_params = WriteParams { max_rows_per_file: 100, ..Default::default()};
+//! let mut dataset = Dataset::write(reader, &uri, Some(write_params)).await.unwrap();
+//! assert_eq!(dataset.get_fragments().len(), 100);
+//!
+//! // Use compact_files() to consolidate the data to 1 fragment
+//! let metrics = compact_files(&mut dataset, Default::default()).await.unwrap();
+//! assert_eq!(metrics.fragments_removed, 100);
+//! assert_eq!(metrics.fragments_added, 1);
+//! assert_eq!(dataset.get_fragments().len(), 1);
+//! # })
+//! ```
+
+use std::collections::HashMap;
+use std::ops::{AddAssign, Range};
+use std::sync::Arc;
+
+use datafusion::physical_plan::SendableRecordBatchStream;
+use futures::{Stream, StreamExt};
+
+use crate::format::Manifest;
+use crate::Result;
+use crate::{format::Fragment, Dataset};
+
+use super::fragment::FileFragment;
+use super::{write_fragments, write_manifest_file, WriteMode, WriteParams};
+
+/// Options to be passed to [compact_files].
+#[derive(Debug, Clone)]
+pub struct CompactionOptions {
+    /// Target number of rows per file. Defaults to 1 million.
```

Comment:
There is one thing that i'd like to do for long time, is that using stats of column to find a better encoding when doing compaction. For example, in Procella paper, it actually do two phase write to determine the optimal perf/cost of encodings, and rewrite the data in background.

Reply:
That's a good point. With that in mind, it means that we need to put page encoding metadata with each page?
Also are there chances that we want to merge/split pages?

Reply:
> Also are there chances that we want to merge/split pages?

This should do that automatically, since it reads the data in as Arrow, and then streams that into the writer.

> With that in mind, it means that we need to put page encoding metadata with each page?

Yup. That's part of the variable encodings design.